### DISCUSSION TITLE
Allow encryption using id-less public keys

### DISCUSSION THREAD

**Discussion by drzraf** [openpgpjs/openpgpjs#1221] [Ideas]
Quoting @nwalfield in https://github.com/openpgpjs/openpgpjs/issues/1144
> Preferences are stored on the addressed User ID's self signature (or the primary User ID, if the certificate is addressed by its key id or fingerprint), but the RFC is explicit that if the preferences are not found on the User ID, then the implementation should fallback to the direct key signature, from Section 2.3.3:

    subpackets on the direct-key signature apply to the entire key.

> As such, it is safe to use a certificate without a User ID. Of course, for the certificate to be valid, it has to then have a direct key signature.

(@wiktor-k)

**Comment by twiss** [comment]
I think we should be somewhat careful with this. Imagine you have the following key:

<details><summary>Sequoia Dump</summary><p>

```
Public-Key Packet, new CTB, 525 bytes
    Version: 4
    Creation time: 2016-07-01 10:00:00 UTC
    Pk algo: RSA (Encrypt or Sign)
    Pk size: 4096 bits
    Fingerprint: 7387 470E 882A F4B6 FEA7  E4B0 EAF8 4862 B87F E3F2
    KeyID: EAF8 4862 B87F E3F2
  
Signature Packet, new CTB, 567 bytes
    Version: 4
    Type: DirectKey
    Pk algo: RSA (Encrypt or Sign)
    Hash algo: SHA1
    Hashed area:
      Signature creation time: 2016-07-01 10:00:00 UTC
      Revocable: true
      Revocation key: BF90 52A4 348F 1464 4D71  D4EC 4211 AAB7 D0A0 3C23/RSA (Encrypt or Sign)
    Unhashed area:
      Issuer: EAF8 4862 B87F E3F2
    Digest prefix: 64B5
    Level: 0 (signature over data)
  
User ID Packet, new CTB, 59 bytes
    Value: "betalarbra7@countermail.com" <betalarbra7@countermail.com>
  
Signature Packet, new CTB, 576 bytes
    Version: 4
    Type: PositiveCertification
    Pk algo: RSA (Encrypt or Sign)
    Hash algo: SHA1
    Hashed area:
      Signature creation time: 2016-07-01 10:00:00 UTC
      Compression preferences: Zlib, Zip
      Hash preferences: SHA256, RipeMD, SHA1
      Symmetric algo preferences: AES256, Twofish, CAST5
      Signature creation time: 2016-07-01 10:00:00 UTC
      Primary User ID: true
      Key flags: CSA
      Issuer: EAF8 4862 B87F E3F2
    Unhashed area:
      Issuer: EAF8 4862 B87F E3F2
    Digest prefix: 8450
    Level: 0 (signature over data)
  
Public-Subkey Packet, new CTB, 525 bytes
    Version: 4
    Creation time: 2016-07-01 10:00:00 UTC
    Pk algo: RSA (Encrypt or Sign)
    Pk size: 4096 bits
    Fingerprint: B347 BE48 1A4D 84AF 05BD  94A8 C501 F067 0B22 481C
    KeyID: C501 F067 0B22 481C
  
Signature Packet, new CTB, 543 bytes
    Version: 4
    Type: SubkeyBinding
    Pk algo: RSA (Encrypt or Sign)
    Hash algo: SHA1
    Hashed area:
      Key flags: EtEr
      Signature creation time: 2016-07-01 10:00:00 UTC
    Unhashed area:
      Issuer: EAF8 4862 B87F E3F2
    Digest prefix: 4FC5
    Level: 0 (signature over data)
```

</p></details>

And imagine that the User ID binding signature failed to parse, due to a bug or unsupported feature in OpenPGP.js, or due to an unknown notation, for example; so that we would skip the packet, and consider the User ID invalid.

The key has a direct key signature, so implementing the proposal here in the most straightforward way would mean that we'd still use the key. However, the direct key signature doesn't specify any preferences, and was only added to set a Revocation Key, as [required by RFC4880bis](https://tools.ietf.org/html/draft-ietf-openpgp-rfc4880bis-08#section-5.2.3.16):

> [The Revocation Key subpacket] is only found on a direct-key self-signature (type 0x1f).  The use on other types of self-signatures is unspecified.

So in that case we would be using the key without its preferences, which is probably not intended.

Of course, then you could say "ok, maybe we could require that the self-signature specifies preferences". But does any key out there actually do that? What's the benefit over requiring a User ID + User ID binding signature?

**Comment by drzraf** [comment]
I'm thinking in a scenario where payload is asymmetrically encrypted, but the identity is managed (published or not) by the website providing the app' (and the key itself). The fact that the encrypted payload does not contains identifiable information (in particular the recipient) can be an advantage in a couple of contexts.
That being said I completely agree that an invalid/unparsable key should trigger a different behavior than a key generated using `--export-filter keep-uid=false`

**Comment by twiss** [comment]
> I'm thinking in a scenario where payload is asymmetrically encrypted, but the identity is managed (published or not) by the website providing the app' (and the key itself). The fact that the encrypted payload does not contains identifiable information (in particular the recipient) can be an advantage in a couple of contexts.

Hmhm, fair enough. I mean, the User ID packet could also say something like "Anonymous", but I agree that it's a bit of a hack.

> That being said I completely agree that an invalid/unparsable key should trigger a different behavior than a key generated using --export-filter keep-uid=false

Yeah, it's just that (at least how the code currently is) these things happen at different levels of abstraction; unparsable packets are skipped during parsing, whereas invalid keys (e.g. without a User ID) are rejected during encryption. So if you solve this the most straightforward way, you would end up with "store the fact that a packet failed to parse, and then later, if there's no UID, reject the key", or something like that, which also seems quite hacky. So I don't really know what a clean way to implement this would be.

**Comment by nwalfield** [comment]
> The key has a direct key signature, so implementing the proposal here in the most straightforward way would mean that we'd still use the key. However, the direct key signature doesn't specify any preferences, and was only added to set a Revocation Key, as [required by RFC4880bis](https://tools.ietf.org/html/draft-ietf-openpgp-rfc4880bis-08#section-5.2.3.16):

This is a good point.  In practice, I'm weary of the preferences.  They are brittle.  When there are multiple recipients, it is hard to choose the best option (what if Alice lists: A, B and Bob lists B, A?).  They are rarely updated.  And, they don't work well in the multi-device (multi-implementation) case.

The RFC says about the [Features](https://tools.ietf.org/html/rfc4880#section-5.2.3.24) subpacket:

> An implementation may freely infer features from other suitable
> implementation-dependent mechanisms.

I think it makes sense to do the same for algorithmic preferences.  In Sequoia, we infer from 2020 that all OpenPGP implementations support SHA-256, SHA-512, AES-128 and AES-256, and use them unconditionally in our mid-level API.

**Comment by nwalfield** [comment]
Also, [the RFC](https://tools.ietf.org/html/rfc4880#section-11.1) does not actually require that a User ID have a self-signature, this is a GnuPG convention:

>     - One Public-Key packet
>
>     - Zero or more revocation signatures
>
>     - One or more User ID packets
>
>     - After each User ID packet, zero or more Signature packets
>       (certifications)
>
>     - Zero or more User Attribute packets
>
>     - After each User Attribute packet, zero or more Signature packets
>       (certifications)

**Comment by wiktor-k** [comment]
> Also, the RFC does not actually require that a User ID have a self-signature, 

Some implementations such as Symantec PGP generated (still generate?) self-sig-less User IDs. IIRC there was a ticket for support of this kind of scenarios for OpenPGP.js

> this is a GnuPG convention:

Isn't this just a way to put prefs? (So for each User-ID a different set of prefs can be used). I don't see any other way (except for direct key sigs that'd be less fine grained... for better or worse :) ).

**Comment by nwalfield** [comment]
> > this is a GnuPG convention:
> 
> Isn't this just a way to put prefs? (So for each User-ID a different set of prefs can be used). I don't see any other way (except for direct key sigs that'd be less fine grained... for better or worse :) ).

The idea of putting the preferences on the User ID was that Alice might have two User IDs: "alice@work.com" and "alice@home.org".  At work, she has a desktop and only checks her work email with MUA #1 and OpenPGP implementation #1.  At home, she has a desktop and only checks her home email with MUA #2 and OpenPGP implementation #2.  When I send Alice a mail, my MUA & OpenPGP implementation look at the identity to find the capabilities of her MUA / OpenPGP implementation.

Today, this setup is atypical.  First, if the two identities are really separate, it would be better to have different certificates.  Second, it is more common to have a MUA configured to check all all emails accounts (home, work, clubs, etc.).  Since the different devices have different capabilities (on my phone, I might use a trusted enclave that supports RSA, and on my desktop a Gnuk that uses ECC), I really want per-device *keys* and the preferences associated with the key.  OpenPGP's preference system doesn't really support this scenario.  As per the RFC, GnuPG falls back to the direct signatures, but it does not, as the RFC leaves open, look for preferences on the subkeys.

**Comment by twiss** [comment]
> I think it makes sense to do the same for algorithmic preferences. In Sequoia, we infer from 2020 that all OpenPGP implementations support SHA-256, SHA-512, AES-128 and AES-256, and use them unconditionally in our mid-level API.

Yeah, this is fair enough. We also assume support of AES-128 and SHA-256 in OpenPGP.js, anticipating RFC4880bis. Still, it may make some sense to keep the preference system, for example in case some weaknesses in SHA2 are theorized, and we all want to switch to SHA3, without waiting for a new spec, and without waiting for all implementations to support it?

> The idea of putting the preferences on the User ID was that Alice might have two User IDs (...) Today, this setup is atypical.

I agree that, if it were designed today, the preference system would be designed very differently. But now that it is as it is, and everyone (I think) puts their preferences in the User ID binding sig, I'm just wondering if it's worth it to push for this to change? OpenPGP.js currently doesn't look for them anywhere else, and I feel like adding that would just make the preference system even more complicated.

**Comment by nwalfield** [comment]
> > I think it makes sense to do the same for algorithmic preferences. In Sequoia, we infer from 2020 that all OpenPGP implementations support SHA-256, SHA-512, AES-128 and AES-256, and use them unconditionally in our mid-level API.
> 
> Yeah, this is fair enough. We also assume support of AES-128 and SHA-256 in OpenPGP.js, anticipating RFC4880bis. Still, it may make some sense to keep the preference system, for example in case some weaknesses in SHA2 are theorized, and we all want to switch to SHA3, without waiting for a new spec, and without waiting for all implementations to support it?

I think it would be faster for implementations to switch their default and for distributions like Debian to ship those changes than to wait for users to update their preferences :D.

> > The idea of putting the preferences on the User ID was that Alice might have two User IDs (...) Today, this setup is atypical.
> 
> I agree that, if it were designed today, the preference system would be designed very differently. But now that it is as it is, and everyone (I think) puts their preferences in the User ID binding sig, I'm just wondering if it's worth it to push for this to change? OpenPGP.js currently doesn't look for them anywhere else, and I feel like adding that would just make the preference system even more complicated.

GnuPG falls back to the direct signature, as described in 4880.  As far as I know no implementation looks for preferences on subkey binding signatures.

**Comment by twiss** [comment]
> I think it would be faster for implementations to switch their default and for distributions like Debian to ship those changes than to wait for users to update their preferences :D.

You may be right. I guess we have a slightly different frame of reference, since e.g. in ProtonMail most of our keys are generated by our own client applications, so then I feel like we can update the preferences in our keys faster than waiting for Debian ^.^

> GnuPG falls back to the direct signature, as described in 4880. As far as I know no implementation looks for preferences on subkey binding signatures.

Hmhm. Are you aware of any implementations that put preferences in the direct key signature?

**Comment by nwalfield** [comment]
In Sequoia we do.  I haven't investigated what others do.

**Comment by drzraf** [comment]
Worth mentioning:
https://keys.openpgp.org/about/faq#older-gnupg
https://dev.gnupg.org/T4393#133689 (About UID-less `update` rather than `UID-key`)
